runtime/onert/core/src/ir/train/TrainableGraph.cc

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/train/TrainableGraph.h"
+
+namespace onert
+{
+namespace ir
+{
+namespace train
+{
+
+TrainableGraph::TrainableGraph(const Graph &graph) : _graph{graph}, _operations{} {}
+
+TrainableGraph::TrainableGraph(Graph &graph, const TrainableOperations &toperations)
+  : _graph{graph}, _operations{}
+{
+  graph.operations().iterate([&](const onert::ir::OperationIndex &idx, onert::ir::Operation &op) {
+    const auto &trainable_op = toperations.at(idx);
+    appendTrainableOperationOnly(idx, trainable_op.clone(op));
+  });
+}
+
+TrainableGraph::~TrainableGraph(void) = default;
+
+OperandIndex TrainableGraph::addOperand(const Shape &shape, const TypeInfo &type)
+{
+  return _graph.addOperand(shape, type);
+}
+
+OperandIndex TrainableGraph::addOperand(OperandIndex index, std::unique_ptr<Operand> &&operand)
+{
+  return _graph.addOperand(index, std::move(operand));
+}
+
+OperationIndex TrainableGraph::addOperation(OperationIndex index,
+                                            std::unique_ptr<ITrainableOperation> &&operation)
+{
+  const auto &inputs = operation->getInputs() | ir::Remove::UNDEFINED | ir::Remove::DUPLICATED;
+  const auto &outputs = operation->getOutputs() | ir::Remove::UNDEFINED | ir::Remove::DUPLICATED;
+  for (auto input : inputs)
+    if (!operands().exist(input))
+      return OperationIndex{};
+  for (auto output : outputs)
+    if (!operands().exist(output))
+      return OperationIndex{};
+
+  auto ind_gen = appendTrainableOperationOnly(index, std::move(operation));
+
+  return ind_gen;
+}
+
+OperationIndex
+TrainableGraph::appendTrainableOperationOnly(OperationIndex index,
+                                             std::unique_ptr<ITrainableOperation> &&operation)
+{
+  auto ind_gen = _operations.push(std::move(operation), index);
+  if (ind_gen.valid())
+  {
+    assert(ind_gen == index);
```

Comment:
I think `assert` failure can happen since `_operations.push` is `unordered_map::emplace` internally.

Reply:
I intended that failure, but looking at the other code again, I see the parts that need to be fixed.